infrastructure/src/main/java/org/corfudb/infrastructure/logreplication/infrastructure/CorfuReplicationDiscoveryService.java

```diff
@@ -857,7 +857,21 @@ public void shutdown() {
         }
 
         if (logReplicationMetadataManager != null) {
-            logReplicationMetadataManager.shutdown();
+            try {
+                IRetry.build(IntervalRetry.class, () -> {
+                    try {
+                        logReplicationMetadataManager.clearLrTables();
```

Comment:
we can invoke this method on a clean shutdown of the service externally by the user or a script.  In that case, LR should restart with the older metadata and not rebuild everything IMO.

Reply:
Also, shutdown() invokes replicationManager.stop() which will now clear the entries

Reply:
I was thinking that if an fsm is shut down, all the data related to its states (which is basically the LR internal tables) should be cleared. And since we anyways create a new FSM if needed, we can start with the new metadata.
Else, with the current design, we cannot delete any metadata anytime.


regarding the shutdown: true, but that is expected.
But I had missed where on lockRelease we call replicationManager.stop(). I'll check a way around that. Thanks for pointing it!

Reply:
By clearing the tables on a shutdown, we make LR stateless in all cases, which does not seem the intended design IMO.  But we can discuss further.

I agree with deleting the data when a cluster is removed because we are sure that it will no longer be required.  One suggestion to decouple this case from the others is to not reuse the "local leader loss" event for this workflow.

Reply:
I have decoupled it in the last push..only removing the entries when a cluster has been removed from the topology now.